### DISCUSSION TITLE
My test suite shows that sending a message to an actor in Kameo with a multithreaded context is slower than in Actix. Am I mistaken?

### DISCUSSION THREAD

**Discussion by xuchaoqian** [tqwewe/kameo#24] [General]
Here is my test suite: [rust-actors-benches](https://github.com/xuchaoqian/rust-actors-benches.git)

# rust-actors-benches

## Preconditions

- CPU: 3.1 GHz Quad-Core Intel Core i7
- Memory: 16 GB 2133 MHz LPDDR3
- MacOS: 13.6.5
- Rust: 1.78
- Actix: 0.13.3
- Ractor: 0.9.7
- Kameo: 0.8.0

## cargo bench --bench actix

- Creation of 100 actors time: **55.924 µs**
- Creation of 10000 actors time: **15.639 ms**
- Waiting on 100000 messages to be processed [ by **single-threaded + do_send** ] time: **16.563 ms**
- Waiting on 100000 messages to be processed [ by **multi-threaded + do_send** ] time: **29.702 ms**
- Waiting on 100000 messages to be processed [ by **single-threaded + send** ] time: **244.69 ms**
- Waiting on 100000 messages to be processed [ by **multi-threaded + send** ] time: **963.50 ms**

## cargo bench --bench ractor

- Creation of 100 actors time: **1.6219 ms**
- Creation of 10000 actors time: **164.34 ms**
- Waiting on 100000 messages to be processed [ by **single-threaded + cast** ] time: **39.842 ms**
- Waiting on 100000 messages to be processed [ by **multi-threaded + cast** ] time: **38.489 ms**
- Waiting on 100000 messages to be processed [ by **single-threaded + call** ] time: **126.76 ms**
- Waiting on 100000 messages to be processed [ by **multi-threaded + call** ] time: **1.2307 s**

## cargo bench --bench kameo

- Creation of 100 actors time: **155.83 µs**
- Creation of 10000 actors time: **17.216 ms**
- Waiting on 100000 messages to be processed [ by **single-threaded + send_async** ] time: **13.898 ms**
- Waiting on 100000 messages to be processed [ by **multi-threaded + send_async** ] time: **22.205 ms**
- Waiting on 100000 messages to be processed [ by **single-threaded + send** ] time: **80.683 ms**
- Waiting on 100000 messages to be processed [ by **multi-threaded + send** ] time: **1.1576 s**

## Comparison

<img width="1238" alt="rust-actors-benches" src="https://github.com/tqwewe/kameo/assets/1064052/216c49b5-e833-429c-9837-9a5aa4d7e202">

**Comment by tqwewe** [comment]
Hi @xuchaoqian, thanks for putting this together! Very nice benchmark. I had a look through the code and didn't see anything wrong with it, so the numbers are probably correct as it stands.

I saw a similar post you made on Ractor's GH, and pretty much the same applies here as what @slawlor said in this comment https://github.com/slawlor/ractor/discussions/209#discussioncomment-8967864

I'm not sure about the internals of Actix, but Kameo also boxes all messages, which might have a small performance impact. Additionally Kameo just performs simple spawning of actors within the tokio runtime, whereas Actix has custom runtime features such as "Arbiters" and such. I personally prefer the simplicity something like Kameo or Ractor brings over Actix.

In terms of messages per second based on this benchmark for multi-threaded, it would be:
- Actix: 9.635 μs
- Ractor: 12.307 μs
- Kameo: 11.576 μs

Though in many of the other benchmarks it seems like Kameo is quite faster, especially in a single-threaded context.

**Reply by xuchaoqian** [reply]
Thank you for your confirmation. I prefer simplicity without the concepts of Arbiters and Systems as well : ).
